File: code/ch02/2.2.1.threeSum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        nums.sort()
        res = []

        # 要找到三个数字，因此只需要找到倒数 n-3 个数字即可
        for i in range(n - 2):
            # 去重
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            # 固定 i，寻找 l 和 r ，使用双指针法
            l = i + 1
            r = n - 1
            while l < r:
                if nums[i] + nums[l] + nums[r] < 0:
                    l += 1
                elif nums[i] + nums[l] + nums[r] > 0:
                    r -= 1
                else:
                    res.append([nums[i], nums[l], nums[r]])
                    # 去重
                    while l < r and nums[l] == nums[l + 1]:
                        l += 1
                    while l < r and nums[r] == nums[r - 1]:
                        r -= 1
                    l += 1
                    r -= 1

        return res

File: code/ch02/2.2.1.threeSum.java
class Solution {
  public List<List<Integer>> threeSum(int[] nums) {
    int n = nums.length;
    Arrays.sort(nums);
    List<List<Integer>> res = new ArrayList<>();

    for (int i = 0; i < n - 2; i++) {
      if (i > 0 && nums[i] == nums[i - 1]) continue;
      int l = i + 1;
      int r = n - 1;
      while (l < r) {
        if (nums[i] + nums[l] + nums[r] < 0) {
          l += 1;
        } else if (nums[i] + nums[l] + nums[r] > 0) {
          r -= 1;
        } else {
          res.add(new ArrayList<Integer>(Arrays.asList(nums[i], nums[l], nums[r])));
          while (l < r && nums[l] == nums[l + 1]) {
            l += 1;
          }
          while (l < r && nums[r] == nums[r - 1]) {
            r -= 1;
          }
          l += 1;
          r -= 1;
        }
      }
    }
    return res;
  }
}
